#469: Add parameterized new support

Nek-12 opened the pull request:
Here's the implementation of the feature proposed in #469. 

It introduces `new()` overload that takes the parameter provided by the factory / multiton binding into the resolution scope. 

The function tries to resolve the parameter and inject it **once** and:

1. If not used, injects it.
2. If used for previous parameter, tries to inject an instance from the DI graph
3. If neither are found, throws `Di.NotFoundException`

This implementation has some quirks that I will mention in comments.

Before merging, they need to be evaluated to align with the maintainers' vision and/or best practices.

closes #469

Nek-12 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.11.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.12.1-bin.zip
```
I upgraded gradle to use jvm 23, hopefully its fine. If not, I can revert the change.

Nek-12 commented on kodein-di/src/commonMain/kotlin/org/kodein/di/bindProvider.kt:
```diff
@@ -35,6 +35,8 @@ public inline fun <reified T: Any> DI.Builder.bindProvider(
+// region bindProviderOf overloads
+
 // TODO This fails with Kotlin/JS Legacy
 //  Re-enable this with IR only target.
```
I'm not sure what this was referring to, but I removed the `noinline` call as it actually was not needed and was affecting performance negatively

romainbsl replied:
Sounds like it is an irrelevant comment, probably forgot to remove it.

Nek-12 commented on kodein-di/src/commonTest/kotlin/org/kodein/di/Tests_28_ParameterizedNew.kt:
```diff
@@ -0,0 +1,107 @@
+        val subclass = S()
+        val b: B by di.instance(arg = subclass)
+
+        assertEquals(b.p, subclass, "injected parameter ${b.p} is not the expected subclass")
```
Right now, this test passes if the parameters are subclasses of the intended type. 

For example, you can see that we injected `S`, a child of `A` and they were injected successfully. This can lead to problems if the client (`B`) didn't actually intend to receive subclasses, but I believe this is reliant on LSP. If the developer follows LSP, then class B won't notice the manipulation. 

This implementation only allows subclasses of A (using `isAssignableFrom`), but a more performant solution would leverage `instanceOf`, allowing any implementation of A, even if an interface. This would reduce safety, but improve construction performance.
I'm not sure how well reflection like this will work here, because the reflection is abstracted under another kosi-libs library. 

Suggestion: Evaluate whether `isAssignableFrom` is actually needed or if it can be replaced with a single `is` check. Choose the more performant option if subclass usage is approved, and opt for `::class` check if not.

romainbsl replied:
Trying to wrap my head around this before I can give a proper answer.

Nek-12 replied:
The Class.isAssignableFrom method is part of the reflection API and is used to determine if instances of one class can be assigned to a variable of another class. Essentially, it checks if the class or interface represented by the current Class object is either the same as, or a superclass or superinterface of, the class or interface represented by the specified Class parameter. It does this at runtime, thus applicable when we have a 

I don't understand this friggin logic myself fully.

TLDR; There will be 2.5% less performance, and we'll have differences in K/N and with inline classes. Additionally, if reflection is involved, things will stop working as intended with `is` check since we check the type provided at compile-time.

But honestly, after spending an hour with chat GPT, SO, and Kotlin playground, I can't find any differences for our use case or reasons why we shouldn't opt into using `is`. Maybe you have more context than me.

romainbsl replied:
the function `isAssignableFrom` you used, comes from [Kaverit](https://github.com/kosi-libs/Kaverit) (Kodein's type system library).

We did develop a specific type system, using `TypeToken`s because Kotlin does not allow much reflection on Non JVM platforms. So heavily using inline functions and reified types we manage to handle more cases than with standard reflection APIs. This is mostly valuable when working with generics, avoiding type erasure.


However, I did try to use `::class` instead of `isAssignableFrom` with something like: 

```kotlin

@PublishedApi
internal interface ParameterizedNew {
    operator fun <T : Any> invoke(t: KClass<T>): T
}

@PublishedApi
internal class DefaultParameterizedNew<A>(
    private val param: A,
    private val context: DirectDI
) : ParameterizedNew {
    private var consumed by atomic(false)

    @Suppress("UNCHECKED_CAST")
    override fun <T : Any> invoke(t: KClass<T>): T {
        if (consumed || !t.isInstance(param))
            return context.Instance(erased(t), null)
        consumed = true
        return param as T
    }
}

@PublishedApi
internal inline fun <reified A : Any, T> DirectDIAware.parameterized(
    param: A,
    block: ParameterizedNew.() -> T,
) = DefaultParameterizedNew(param, directDI).run(block)

@PublishedApi
internal inline fun <reified T : Any> ParameterizedNew.invoke(): T = invoke(T::class)
```

But, the result remains the same, it doesn't solve the issue for Non JVM platforms. 

So, at the moment, I suggest we leave this as is, using Kaverit type system to maintain coherence and full capacity at runtime.

However, subclass testing works on the JVM so we can move this test to `jvmTest` target. and eventually add documentation accordingly.

Nek-12 replied:
The T::class would restrict the ability to match arguments to **exactly type T**, no subclasses, no upperclasses. That is the main difference. 

So you're saying we just leave the `isAssignableFrom` check? I'm curious as to why an `is` check didn't solve the problem for you since it's a Kotlin Compiler feature.

romainbsl replied:
> The T::class would restrict the ability to match arguments to exactly type T, no subclasses, no upperclasses. That is the main difference.

True!

> I'm curious as to why an is check didn't solve the problem for you since it's a Kotlin Compiler feature.

It is, but it does not solve key handling for the container, TypeToken's are mostly used to manipulate / explore the DI container tree. It helps us keeping metadata about types at compile time and runtime. In that regards, `is` does not helps much.

> So you're saying we just leave the isAssignableFrom check?

I do understand the point of having some mechanism using `is` here, but I haven't a much better implementation to suggest ATM. So the `isAssignableFrom` is the least we can do know, before being able to do something better. It is internal, though, it is armless for the users. 

Is that ok for you ? 

I give myself a week or two to try to figure something without the `isAssignableFrom` check.

Nek-12 replied:
Yes, we can use the `isAssignableFrom`

Nek-12 replied:
Hey, I just found something. The java/kotlin STDLib on JVM has the `isInstance` function, which is equivalent to `is` but for `KClass` instances. In the future, consider adding it to org.kodein.type API and using here, as it should be more performant than using isAssignableFrom

romainbsl replied:
`isInstance` allows to check if an instance is of a given type, not to compare types between them, which is the main goal of Kodein's types.

Nek-12 commented on kodein-di/src/commonTest/kotlin/org/kodein/di/Tests_28_ParameterizedNew.kt:
```diff
@@ -0,0 +1,107 @@
+            bind { factory { p: A -> new(p, ::D) } }
+        }
+
+        val param = A() // injected class does not use the param
```
Here, this test passes even though we never used the parameter at all. It was not used in the injected dependency (`D`) constructor. 

So the parameter `A` that the client provided is wasted.

This can lead to unintended behavior if the developer made a mistake of not including parameters, but accidentally used another class, which happened to also be present in the DI graph. 

However, it's not possible to fix this in compile time, and the only solution I see is to throw something like `DI.UnusedParameterException` after creating an instance.

Suggestion: Choose one option: 

1. Leave as is : more convenient and lenient, but can cause runtime bugs
2. Change to add verification code and throw: Strict, but will crash in runtime at instance's first usage.

romainbsl replied:
I'd rather see something more strict. I'd prefer have a runtime exception about an unused parameter, meaning my code is not properly designed, than having wired behavior from product perspective.

So let's go for 2. if you're ok with that.

Nek-12 commented on kodein-di/src/commonTest/kotlin/org/kodein/di/Tests_28_ParameterizedNew.kt:
```diff
@@ -0,0 +1,107 @@
+    @Test
+    fun test_03_new_bindMultitonOf_param() {
+        val di = DI {
+            bindMultitonOf<A, _, _>(::B)
```
You can see that even though we added overloads `bindMultitonOf` and `bindFactoryOf`, we cannot not provide the type of the argument, as there is nothing indicating what it is. 

This leads to us writing boilerplate like `<ParameterType, _, _, _, _, _, _, _>`. 

Kotlin does not yet have syntax for partial type arg specification and there is no lambda to define the type of the param. This also leads to compilation errors each time a parameter is added or removed.

Suggestion: Leave, but document that using 
`bind* { it: ParameterType -> new(it, ::Class)` }` is a more convenient solution, 

OR

Remove all `bindMultitonOf` / `bindFactoryOf` overloads from this PR.

romainbsl replied:
As a user, I really don't like the `<ParameterType, _, _, _, _, _, _, _>` boilerplate...

The parameterized new is already a nice improvement, avoiding chained `instance()...` calls, don't you think?

Nek-12 replied:
Yes I think 99% of people will use the `new` op. The question is only whether we should leave the shitty overloads just in case someone decides to use them or we manage to improve them in the future.

romainbsl replied:
I would go against those overloads, if someone needs it it can still be done in there code base. And if we found something to do this properly, we should implement a that time.

Nek-12 commented on kodein-di/src/commonMain/kotlin/org/kodein/di/New.kt:
```diff
@@ -129,4 +129,521 @@ public inline fun <T,
+/**
+ * @see new
+ */
+public inline fun <
```
I added more overloads of `new` and other functions because 22 is the maximum value actually supported by Kotlin, with 2 proofs:

1. 22 parameters is the maximum number for which Kotlin generates `componentN()` functions on data classes. 
2. Kotlin supports referring to functions and creating their instances. `Function22` is the highest supported number of parameters for function interface references.

I did this because 10 parameters are actually a relatively common classes like ViewModels, which can inject a lot of use-cases etc. When the `new` function breaks at 10 params, it quickly becomes annoying to completely rewrite the declaration to use `instance()` everywhere. 

I get it's a huge pile of boilerplate, but there is no other way to do this at the moment as vararg type arguments are not implemented in kotlin.

Nek-12 commented:
Additionally, looks like `isAssignableFrom` does not properly work and fails the test on macOS. 

```
test_02_new_subclass[macosArm64]: 

org.kodein.di.DI.NotFoundException: No binding found for B, with argument S
```

This is likely a bug in `org.kodein.type` related to how native code gets executed. I've seen some tickets on YT re: usage of `kotlin.reflect.typeOf` on Native.

romainbsl commented:
Thanks a lot for your contribution!
I'll need few days to validate, but sounds great.

Looks like `test_02_new_subclass` is failing on all NON JVM targets, So I'd to check what's going on with kodein types.

Nek-12 commented:
Hi @romainbsl! Did you have some time to check out this PR? I assume the main blocker here is the kodein.type library support. I think for now we could replace the subclass check with an `is` check and maybe make any adjustments later, but what about the rest of the decisions to make?

romainbsl commented:
> Hi @romainbsl! Did you have some time to check out this PR? I assume the main blocker here is the kodein.type library support. I think for now we could replace the subclass check with an `is` check and maybe make any adjustments later, but what about the rest of the decisions to make?

Unfortunately, not so much. 
I'd like to try out something with kodein's type system, but once again I am afraid that we will be limited by the native/js lack of reflection. 
At best, we could merge and document the use of bindings without inheritance.

Nek-12 commented:
Well before we can merge, I just need you, @romainbsl, to make decisions on the rest of the API design. Please respond to comments with your desired approach, and I'll change my code and update the documentation.

Nek-12 commented:
So since all of the comments are addressed, I will make adjustments and update documentation. Feel free to make any adjustments to it on your own, through comments, or otherwise. I'll try to manage the updates within 3 days.

Nek-12 commented:
@romainbsl I pushed commits that addressed all of the issues above and added a new section to the paragraph. I'm not familiar with asciidoc, so please point out any mistakes. Also since one-argumen constructors are now type-safe, I made the `new()` function for one-parameter classes compile-type-safe (the only parameter is the injected argument by law of elimination)

romainbsl commented:
I should be able to review this weekend, I had a hell of a week...

romainbsl commented:
LGTM 👍 
I didn't find anything good using `is` on native platforms, we're always loosing the knowledge on inheritance somehow. And JavaScript will never be working as there is no inheritance there.

romainbsl commented on kodein-di/src/commonTest/kotlin/org/kodein/di/Tests_28_ParameterizedNew.kt:
```diff
@@ -0,0 +1,78 @@
+
+    private class I
+    @Suppress("unused")
+    private class D(val p: I, val p2: A)
```
Using type `A` as second parameter makes the parameter consumed, so the injection succeeds.
```suggestion
    private class D(val p: I, val p2: I)
```

romainbsl commented:
Probably going to release this one in a minor version as the Compose PR might go into a major one regarding the impact.
